server/src/main/java/org/elasticsearch/common/Rounding.java

```diff
@@ -401,8 +404,22 @@ private LocalDateTime truncateLocalDateTime(LocalDateTime localDateTime) {
             }
         }
 
+        /**
+         * Time zones with two midnights get "funny" non-continuous rounding
+         * that isn't compatible with the pre-computed array rounding.
+         */
+        private static final Set<String> HAS_TWO_MIDNIGHTS = Set.of("America/Moncton", "America/St_Johns", "Canada/Newfoundland");
```

Comment:
would it be doable to detect timezones that don't work with this optimization at runtime instead of maintaining an allowlist?

Reply:
I can't think of a way to do right now. At least, not a god way.

Reply:
I thought Canada was changing time at 2 am, on another side, some other known time zones such as `Asia/Gaza` for example seems to indeed have 2 mightnights. I did some quick and dirty test and I have got a slightly different list for timezones in my JVM:
```
America/Asuncion 2020-03-22T00:00 -> 2020-03-21T23:00
America/Havana 2020-11-01T01:00 -> 2020-11-01T00:00
America/Santiago 2020-04-05T00:00 -> 2020-04-04T23:00
America/Scoresbysund 2020-10-25T01:00 -> 2020-10-25T00:00
Asia/Amman 2020-10-30T01:00 -> 2020-10-30T00:00
Asia/Beirut 2020-10-25T00:00 -> 2020-10-24T23:00
Asia/Damascus 2020-10-30T00:00 -> 2020-10-29T23:00
Asia/Gaza 2020-10-31T01:00 -> 2020-10-31T00:00
Asia/Hebron 2020-10-31T01:00 -> 2020-10-31T00:00
Asia/Tehran 2020-09-21T00:00 -> 2020-09-20T23:00
Atlantic/Azores 2020-10-25T01:00 -> 2020-10-25T00:00
Chile/Continental 2020-04-05T00:00 -> 2020-04-04T23:00
Cuba 2020-11-01T01:00 -> 2020-11-01T00:00
Iran 2020-09-21T00:00 -> 2020-09-20T23:00
```

Reply:
Very nice!

Reply:
Is this really only a problem when there are two midnights, or is it a more general problem when the time goes back due to daylight savings, and you just need special values of `offset` to make the bug occur if the transition is not around midnight?

Reply:
@nik9000 please correct me if I am wrong, but the way I understand it, it is the way we truncate it to midnight that is causing the issues, I played with it a bit more and it looks like `atStartOfDay` will truncate `2020-03-22T00:00` to 22nd, but it will truncate the moment after it to 21st, which can cause all sorts of issues. For example:

```
        // America/Asuncion 2020-03-22T00:00 -> 2020-03-21T23:00
        TimeZone tz = TimeZone.getTimeZone("America/Asuncion");
        ZoneRules rules = tz.toZoneId().getRules();
        ZoneOffsetTransitionRule fallback = rules.getTransitionRules().get(0);
        ZoneOffsetTransition transition = fallback.createTransition(2020);
        System.out.println("Before: " + transition.getDateTimeBefore().toLocalDate().atStartOfDay());
        System.out.println("After:  " + transition.getDateTimeAfter().toLocalDate().atStartOfDay());
```
will give us
```
Before: 2020-03-22T00:00
After:  2020-03-21T00:00
```

The same thing for `America/New_York` will return
```
Before: 2020-03-08T00:00
After:  2020-03-08T00:00
```

Reply:
> the way I understand it, it is the way we truncate it to midnight that is causing the issues, I played with it a bit more and it looks like `atStartOfDay` will truncate `2020-03-22T00:00` to 22nd, but it will truncate the moment after it to 21st, which can cause all sorts of issues.

This is pretty much what I saw, yes. In the one time zone I dug into deeply I saw a single instant gets rounded a day ahead of all of the instants around it. Its *weird* but it is what we have. If we didn't do that then we could use the array based rounding everywhere.

I'll dig more tomorrow into your examples. The ones I found I did by randomized testing and the one that I dug into deeply was a problem with having two midnights. I should be able to answer much more precisely what the problem is when I have a closer look.




> Is this really only a problem when there are two midnights, or is it a more general problem when the time goes back due to daylight savings, and you just need special values of `offset` to make the bug occur if the transition is not around midnight?

`offset` is pretty much incompatible with daylight savings time. At least, it doesn't work as advertised. See #56305 for some investigation.

Reply:
Ok! I think it isn't that they have two midnights, it is that transitioned a minute after midnight:
```
Transition[Gap at 2006-04-02T00:01-03:30 to -02:30]
```

This causes rounding to output:
```
1162088880000  1162002600000
1162088940000  1162002600000
1162089000000  1162089000000   <---- This minute is in "tomorrow"
1162089060000  1162002600000   <----- But this minute is in "yesterday"!
1162089120000  1162002600000
1162089180000  1162002600000
1162089240000  1162002600000
1162089300000  1162002600000
```

It is this kind of "forwards and backwards" dance that my simple array based rounding can't handle. I'll see if I can detect it.

Reply:
Even if offsets are not working as advertised with DST, I worry that they would behave differently again with this optimization. It would be safer to disable this optimization whenever there are transitions that make the time go backwards in the local time in the time range of the index?

Reply:
I've pushed a change that disables the optimization if the range falls in one of the times when DST drags you back to "yesterday". And I've pushed a test that checks those times.

As it stands now, I don't think the `offset` feature of date rounding will effect this optimization.

Reply:
I've played with offset to double check and can't cause issues with it. However do you think we could detect timezones that don't work dynamically instead of relying on a static list? E.g. could we iterate transitions in the considered interval and check whether there's one that brings us to a different day?

Reply:
We could use something like the test case uses to find candidates, but it'd require loading all of the time zone rules on startup. I'm hoping that the test can prevent us from having to do that by being very careful about this list.

Reply:
Actually I wasn't thinking about testing all timezones up-front on startup, I was more thinking of doing the test when building the Rounding object by only looking at transitions of the considered timezone.

Reply:
Hmmmm - the `assert` that I have below sort of asserts that. But it isn't nearly as strong. I'll think about it!

Reply:
@jpountz I've pushed a change that does this.